src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs

```diff
@@ -77,26 +77,37 @@ public async Task ExecuteAsync(ExportJobRecord exportJobRecord, WeakETag weakETa
 
             try
             {
-                // Get destination type from secret store.
-                DestinationInfo destinationInfo = await GetDestinationInfo(cancellationToken);
+                // Get destination type from secret store and connect to the destination using appropriate client.
+                await GetDestinationInfoAndConnect(cancellationToken);
 
-                // Connect to the destination using appropriate client.
-                _exportDestinationClient = _exportDestinationClientFactory.Create(destinationInfo.DestinationType);
+                // Update our internal mapping with list of files that we have already committed (when resuming export jobs)
+                // or an empty dictionary (for new export jobs). As we keep modifying our internal mapping, the Output field
+                // will also get updated. These changes will be committed to the data store whenever we update our ExportJobRecord.
+                _resourceTypeToFileInfoMapping = _exportJobRecord.Output;
 
-                await _exportDestinationClient.ConnectAsync(destinationInfo.DestinationConnectionString, cancellationToken, _exportJobRecord.Id);
+                // If we are resuming a job, we can detect that by checking the progress info from the job record.
+                // If it is null, then we know we are processing a new job.
+                if (_exportJobRecord.Progress != null)
+                {
+                    List<Uri> fileUris = _resourceTypeToFileInfoMapping.Values.Select(fileInfo => fileInfo.FileUri).ToList();
+                    await _exportDestinationClient.OpenFilesAsync(fileUris, cancellationToken);
 
-                // TODO: For now, always restart from the beginning. We will support resume in another work item.
-                _exportJobRecord.Progress = new ExportJobProgress(continuationToken: null, page: 0);
+                    _logger.LogTrace("Finished opening existing files for resuming export.");
```

Comment:
I think there was some miscommunication here. I don't think we need to open all of the files that's in progress. You could just load the file when you need it. 

e.g.,

in ProcessSearchResults

1. Check to see if there is a mapping in _resourceTypeToFileInfoMapping.
2. If there is mapping, use it
3. If there is no mapping, then check to see if there is a file or not from progress and open file if there is a file or create file is there isn't one.

This way it doesn't require you to open all files and download all block list up front. Only when it's necessary.

Reply:
Updated the code accordingly.